### DISCUSSION TITLE
WASM Guidance

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#2218] [Q&A]
**TL;DR**
Is there a checklist or guidance for prepping a Server Hosted Oqtane app to be hosted in Web Assembly?

We've built a great Oqtane MVP site that has some mobile capabilities.  It's working fine as a server app, but it's CPU intensive at times and instead of scaling-up the Azure App Service to handle the load, we are considering deploying the site as WASM and keeping the server side as file serving and API.  I went into my site setting and changed the hosting model to Web Assembly.  That broke the site with a dependency not found for a 3rd party library.  I think we need to rethink our packaging a bit to accommodate WASM.  I've started adding client dependencies in the IModule.ModuleDefinition.Dependencies and was able to resolve most of those.  Now I'm seeing what look like server dependencies showing up as errors in the browser console.   I'm about to add every DLL in every bin to the client folder bin and then create dependencies for all of them, but I'm just winging it.

``` 
System.AggregateException: One or more errors occurred. (Could not load file or assembly 'Microsoft.Extensions.Identity.Core, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' or one of its dependencies.)
 ---> System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Extensions.Identity.Core, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' or one of its dependencies.
File name: 'Microsoft.Extensions.Identity.Core, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at System.Reflection.AssemblyExtensions.GetTypes(Assembly assembly, Type interfaceType) in D:\Repos\94-prop-aa\src\oqtane.framework\Oqtane.Shared\Extensions\AssemblyExtensions.cs:line 38
   at System.Reflection.AssemblyExtensions.GetInterfaces[IService](Assembly assembly) in D:\Repos\94-prop-aa\src\oqtane.framework\Oqtane.Shared\Extensions\AssemblyExtensions.cs:line 23
   at Oqtane.Client.Program.RegisterModuleServices(Assembly assembly, IServiceCollection services) in D:\Repos\94-prop-aa\src\oqtane.framework\Oqtane.Client\Program.cs:line 117
   at Oqtane.Client.Program.Main(String[] args) in D:\Repos\94-prop-aa\src\oqtane.framework\Oqtane.Client\Program.cs:line 51
   --- End of inner exception stack trace ---
callEntryPoint @ blazor.webassembly.js:1
await in callEntryPoint (async)
At @ blazor.webassembly.js:1
await in At (async)
(anonymous) @ blazor.webassembly.js:1
(anonymous) @ blazor.webassembly.js:1 ```

**Comment by sbwalker** [comment]
Unfortunately there is no easy way to track down dependencies when running on Blazor WebAssembly. I submitted some feedback to Microsoft about this exact issue almost a year ago. The problem is that the .NET linker tries to be smart and do "tree shaking" to reduce the number of DLLs sent to the browser. However the linker is not able to recognize dynamic dependencies - only static dependencies - which does not work well in a dynamic framework like Oqtane. I should probably post a blog with all of my own research and findings in this area.

**Reply by markdav-is** [reply]
if the answer now to just include everything in client an server bin?  maybe not that simple?

**Reply by sbwalker** [reply]
Not that simple. The linker removes classes and methods from the core (System.*) libraries that it thinks the application isn’t using. The only way to make this work with trimming is to pass all the assemblies that will be loaded into the app into the linker. Every extension library must be seen by the linker. And in .NET 5/6, you also need to use the assemblies outputted by the linker. In .NET 5/6, the linker rewrites the assemblies so they no longer reference “façade” assemblies, like System.Runtime. And then it trims those façade assemblies from the app’s output. So configuring this is complicated - there are a number of config switches for different purposes.

**Reply by markdav-is** [reply]
That's a lot to consider.  So WASM is supported in Oqtane, but it's not easy.  It sounds like I'd need to create a custom build script that would do these extra steps.  Then we'd be sending the entire framework just in case.  We use a mono-module approach with many modules in the same solution.  That mega-module has a dependency on Oqtane for builds.  I'm curious if that megaassembly would have most of what it needed.  This conversation is at a lower level of the build process than I'm comfortable.  I'll take a close look at WASM and the build details so I can familiarize myself with more details.

**Reply by sbwalker** [reply]
This link contains some useful information - https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/configure-linker. The challenge is that during development you are usually working in Debug mode - and in this configuration your application will work just fine when running on WebAssembly. However when you build in Release mode for a deployment, the linker will become active and will trim your output based on its static code/dependency analysis - which usually causes problems that are not immediately obvious until you do a full run-time regression test of your application and see random errors in your browser console. The most foolproof way to handle this is to disable the Linker. However if you do this, you will not get any of the performance benefits of trimming (ie. the DLLs sent from the server to the browser will be larger in size). So you need to make your own choice on the priorities for your app - either accept the unpredictable behavior of the Linker and the trial/error remediation steps to address the WebAssembly run-time issues, or disable the Linker and accept the larger application payload. My personal opinion based on the effort I went through to get the Blog module (which is not overly complicated) to run on WebAssembly, is that most developers will choose predictability over performance.

**Reply by Sen-Gupta** [reply]
Seems like, I'm in time to jump on the train!

So I understand the issues being caused above. Interestingly  we have a case, where I  created a moulde and the shared dll is not being loaded? What am I missing here?

![image](https://user-images.githubusercontent.com/8068582/171687211-6c24e1fd-d2da-4cd4-bbc9-387b9a3e5e3d.png)

The shared dll has only one model file but leverage Generics.
![image](https://user-images.githubusercontent.com/8068582/171691433-02bbd24a-4508-4303-ac9c-8246b885409d.png)

So are we missing dependencies here?

Thanks in advance!

**Comment by sbwalker** [comment]
@Sen-Gupta the IModule interface has a Dependencies property. This property is used to identity any DLLs which the module depends upon that need to be downloaded to the client (when running on WebAssembly):

```
using Oqtane.Models;
using Oqtane.Modules;

namespace Oqtane.Blogs
{
    public class ModuleInfo : IModule
    {
        public ModuleDefinition ModuleDefinition => new ModuleDefinition
        {
            Name = "Blog",
            Description = "Blog",
            Version = "1.0.4",
            ServerManagerType = "Oqtane.Blogs.Manager.BlogManager, Oqtane.Blogs.Server.Oqtane",
            ReleaseVersions = "1.0.0,1.0.1,1.0.3,1.0.4",
            Dependencies = "Oqtane.Blogs.Shared.Oqtane",
            SettingsType = "Oqtane.Blogs.Settings, Oqtane.Blogs.Client.Oqtane"
        };
    }
}
```
The interface above indicates that Oqtane.Blogs.Shared.Oqtane.dll needs to be downloaded to the client If there are multiple dependencies they can be separated by commas.

**Comment by Sen-Gupta** [comment]
Thank you so much. I just can't believe myself. I completely forgot this part.
